src/Microsoft.AspNetCore.Razor.Tools/GenerateCommand.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.VisualStudio.LanguageServices.Razor;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.Tools
+{
+    internal class GenerateCommand : CommandBase
+    {
+        public GenerateCommand(Application parent)
+            : base(parent, "generate")
+        {
+            Sources = Argument("sources", ".cshtml files to compile", multipleValues: true);
+            ProjectDirectory = Option("-p", "project root directory", CommandOptionType.SingleValue);
+            OutputDirectory = Option("-o", "output directory", CommandOptionType.SingleValue);
+            TagHelperManifest = Option("-t", "tag helper manifest file", CommandOptionType.SingleValue);
+        }
+
+        public CommandArgument Sources { get; }
+
+        public CommandOption OutputDirectory { get; }
+
+        public CommandOption ProjectDirectory { get; }
+
+        public CommandOption TagHelperManifest { get; }
+
+        protected override Task<int> ExecuteCoreAsync()
+        {
+            var result = ExecuteCore(
+                projectDirectory: ProjectDirectory.Value() ?? Environment.CurrentDirectory,
+                outputDirectory: OutputDirectory.Value(),
+                tagHelperManifest: TagHelperManifest.Value(),
+                sources: Sources.Values.ToArray());
+
+            return Task.FromResult(result);
+        }
+
+        protected override bool ValidateArguments()
+        {
+            if (string.IsNullOrEmpty(OutputDirectory.Value()))
+            {
+                Error.WriteLine($"{OutputDirectory.ValueName} not specified.");
+                return false;
+            }
+
+            if (Sources.Values.Count == 0)
+            {
+                Error.WriteLine($"{Sources.Name} should have at least one value.");
+                return false;
+            }
+
+            return true;
+        }
+
+        private int ExecuteCore(string projectDirectory, string outputDirectory, string tagHelperManifest, string[] sources)
+        {
+            tagHelperManifest = Path.Combine(projectDirectory, tagHelperManifest);
+            outputDirectory = Path.Combine(projectDirectory, outputDirectory);
+
+            var tagHelpers = GetTagHelpers(tagHelperManifest);
+
+            var engine = RazorEngine.Create(b =>
+            {
+                RazorExtensions.Register(b);
+
+                b.Features.Add(new StaticTagHelperFeature() { TagHelpers = tagHelpers, });
+            });
+
+            var templateEngine = new MvcRazorTemplateEngine(engine, RazorProject.Create(projectDirectory));
```

Comment:
We'll eventually need to abstract this to support multiple versions of Razor.

Reply:
Added to the checklist 👍

Reply:
Yep. This data will need to flow through the SDK